GH-137759:Limit _PyObject_HashFast to dicts keys, rename it, and mark it as Py_ALWAYS_INLINE

JasonMendoza2008 opened the pull request:
Context:

`setobject.c` used a mix of `_PyObject_HashFast` and `PyObject_Hash`. The consensus[^1] was that `_PyObject_HashFast` should only really be used for dict keys (string-only sets are not so omnipresent). This PR uses the standard `PyObject_Hash` without the fast path (making `set[Any]` faster and `set[str]` slower[^2]), renames `_PyObject_HashFast` since its name was misleading and marked it as `Py_ALWAYS_INLINE` to force inlining.

Pretty sure this doesn't require a NEWS entry?


[^1]: There were discussions in the same issue around removing `_PyObject_HashFast` completely and adding the fast path for strings (and even potentially integers) to `PyObject_Hash` but there was no consensus so I'm not including it for now.
[^2]: In this [comment ](https://github.com/python/cpython/pull/137828#issuecomment-3219968888), I run `pyperformance` on two versions of `setobject.c`, one using HashFast everywhere and one using the standard Hash everywhere. None of these correspond to the current implementation of `setobject.c` since the current implementation uses a mix of both. Obviously in all cases, dictionaries use HashFast.


<!-- gh-issue-number: gh-137759 -->
* Issue: gh-137759
<!-- /gh-issue-number -->

ZeroIntensity commented:
> making set[Any] faster and set[str] slower

I'm not totally convinced that this is true. Detecting the fast-path for strings is just one pointer comparison, which really won't have a noticeable performance impact. I agree that we want `_PyObject_HashFast`/`_PyObject_HashDictKey` for inlining purposes, but I don't see why the fast path has to be limited to it.

ZeroIntensity commented:
> but then do we even need HashFast, we could just put everything in PyObject_Hash

Yeah, because even if there's no speedup there, it's just simpler maintenance (in the future, we might want to make other assumptions about a dictionary key's hash that improve performance).

JasonMendoza2008 commented:
> I'm not totally convinced that this is true. Detecting the fast-path for strings is just one pointer comparison, which really won't have a noticeable performance impact

But if we don't detect it, we're a a function pointer dereference and call away from getting the hash anyway, right? The hash is not recomputed without that fast-path since it's stored

JasonMendoza2008 commented:
> > making set[Any] faster and set[str] slower
> 
> I'm not totally convinced that this is true. Detecting the fast-path for strings is just one pointer comparison, which really won't have a noticeable performance impact. I agree that we want `_PyObject_HashFast`/`_PyObject_HashDictKey` for inlining purposes, but I don't see why the fast path has to be limited to it.

Here is a comparison using pyperformance for what it's worth.

```
Pyperformance version: 1.11.0
Report on Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.39
Number of logical CPUs: 24
```

<details>
<summary>Run 1</summary>

<img width="989" height="1990" alt="image" src="https://github.com/user-attachments/assets/9f149e9d-3dad-4a1e-974b-9de3e88eefd8" />

</details>

<details>
<summary>Run 2 (using the -r option which is supposed to get more accurate results by running everything more)</summary>
<img width="989" height="1990" alt="image" src="https://github.com/user-attachments/assets/3fc0af95-fc6c-4e7e-ab2f-c1054829e845" />
</details>

Both versions were compiled with `./configure --enable-optimizations --with-lto`

JasonMendoza2008 commented:
> Using `_PyObject_HashFast()` in `set_intersection()` sounds reasonable to me.

Should someone that has the authority to do so do that? Not sure how to move forward this PR.

MojoVampire commented:
I think getting `_PyObject_HashFast` out of everything but the `dict` code makes sense. It's an optimization there because `str`-keyed `dict`s are a *massive* percentage of all `dict`s (simply because they implement all Python namespaces under-the-hood). If it's used *anywhere* else, we may as well just make `PyObject_Hash` do it itself, because we've decided that `str`s are the primary type to be used for hashing everywhere, for any reason. I just can't see any scenario in which `dict` *and* `set` should use it, but nothing else.

JasonMendoza2008 commented:
> LGTM. Please resolve conflicts.

Will do asap

JasonMendoza2008 committed:
GH-137759: Limit _PyObject_HashFast to dict keys

JasonMendoza2008 committed:
GH-137759: Limit _PyObject_HashFast to dict keys

JasonMendoza2008 commented:
@serhiy-storchaka


I think everything is ok? I squashed everything into one commit and applied the diff as one commit on main. New to this, so idk if that was done correctly, but it looks like it was.


EDIT: looks like something is broken, my bad - not sure what it is though, it builds fine on my computer

serhiy-storchaka commented:
@vstinner, besides a specialized hash function, dicts has special code path for string keys which optimizes comparison. Dicts with string-only keys are _extremely_ common. Sets of strings... even if they are relatively common, they are not more common than sets of other types combined.

@rhettinger, are you fine with this change?

vstinner commented:
If we make this change, we should carefully record the decision (for future readers). For example, in _PyObject_HashFast() comment. Explain why it's used on dict but not on other types such as set.

serhiy-storchaka committed:
Apply suggestion from @serhiy-storchaka

serhiy-storchaka committed:
Merge branch 'main' into fix-issue-137759

JasonMendoza2008 commented:
Thanks @serhiy-storchaka, I meant to add the comment but forgot.

JasonMendoza2008 commented:
Huh didn't mean to ping that many people, was just squashing the commits, sorry if that wasn't supposed to happen.
EDIT: Yeah I think I messed up trying to squash the commits, it now says I changed 252 files.
EDIT2: ok I fixed it, it's squashed properly now, sorry for the global ping, I don't really know how to remove the auto request reviews for everyone not involved. 
EDIT3: Thanks StanFromIreland

hugovk commented:
For next time, please don't force push: https://devguide.python.org/getting-started/pull-request-lifecycle/#don-t-force-push

vstinner commented:
Merged, thank you.

JasonMendoza2008 commented:
> For next time, please don't force push: https://devguide.python.org/getting-started/pull-request-lifecycle/#don-t-force-push 

Oh ok, noted. Thank you.
